codegen/ace/src/layout.rs

```diff
@@ -0,0 +1,230 @@
+use crate::circuit::Node;
+use air_ir::{Air, Identifier, PublicInputAccess, TraceAccess};
+use std::collections::BTreeMap;
+use std::ops::Range;
+
+/// For each set of inputs read from the transcript, we treat them as extension field elements
+/// and pad them with zeros to the next multiple of 4. They can then be unhashed to a double-word
+/// aligned region in memory.
+const HASH_ALIGNMENT: usize = 4;
+
+const NUM_QUOTIENT_PARTS: usize = 8;
+
+/// Describes the layout of inputs given to an ACE circuit.
+/// Each set of variables is aligned to the next multiple of 4, ensuring they can be efficiently
+/// unhashed from the transcript and that each input region is aligned to [`HASH_ALIGNMENT`].
+///
+/// We assume the following about the underlying `Air` from which the layout is constructed
+/// - The proof always contains a `main` and `aux` segment, even when the latter is unused,
+/// - The maximal degree of an [`Air`] is `9`, such that the quotient can be decomposed in 8 chunks.
+///   TODO(@adr1anh): Derive the degree generically.  
+#[derive(Clone, Debug, Default, Eq, PartialEq)]
+pub struct Layout {
+    /// Region for each set of public inputs, sorted by `Identifier`
+    pub public_inputs: BTreeMap<Identifier, InputRegion>,
+    /// Region for auxiliary random inputs.
+    pub random_values: InputRegion,
+    /// Regions containing the evaluations of each segment, ordered by `trace[row_offset][segment]`.
+    ///
+    /// # Detail:
+    /// Note that we make the following assumptions which do not affect the evaluation of the
+    /// circuit but facilitate the implementation of the overall MASM verifier. In particular,
+    /// these properties facilitate the computation of the DEEP composition polynomial and
+    /// can lead to more efficient unhashing from the transcript into memory.
+    /// - The [`Air`] from which the layout is derived can only contain a *main* and *aux* trace
+    ///   (the latter can be empty).
+    /// - We treat the *quotient* as the third trace, handling it in the same way as the witness
+    ///   traces. This requires the prover to provide the out-of-domain evaluations in
+    ///   the *next* row of each quotient part. These are unused by the circuit.
+    /// - The rows must be ordered as follows:
+    ///   ```ignore
+    ///   main_curr, aux_curr, quotient_curr, main_next, aux_next, quotient_next.
+    ///   ```
+    /// - Each trace must be padded with zero columns such that each row is word-aligned.
+    ///
+    /// # TODO(adr1anh):
+    /// The degree of the quotient is fixed to 8 matching the degree of the VM constraints, but
+    /// the actual degree can be derived from the [`Air`].
```

Comment:
If we have opened an issue for this, then we can probably just remove this, unless you have a strong preference for keeping it

Reply:
#391